### PR TITLE
Improve templates integration test

### PR DESCRIPTION
**Pull request by mariaschuld**
**Context:**

When adding templates (PR #497) I noticed that the integration tests make the implicit assumption that differentiable arguments are always kwargs to a template.

This is a refactor of the templates integration test to make it more flexible, elegant and logical.

**Description of the Change:**

The template tests now take as arguments the template and a dictionary of differentiable inputs, and a dictionary of non-differentiable inputs. This way the user can define which inputs to test as a positional argument to the qnode. 

Other improvements were: 
* Simplify the initialization integration tests
* Make the _get_shape input check more rigid, since it pulled the wrong shape for tensorflow variables
* Make names consistent and avoid abbreviations

**Benefits:** The test should be much clearer and more powerful now, getting rid of the confusing use of "hyperparameters", "inpts", "args" and "kwargs". 

**Possible Drawbacks:** The tests are rather complex to read, but have the advantage that a contributor of a new template only has to add one line to a list.

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/templates/utils.py [hunk 1]
```diff
@@ -76,9 +77,26 @@ def _get_shape(inpt):
     Returns:
         tuple: shape of ``inpt``
     """
-    inpt = np.array(inpt)
 
-    return inpt.shape
+    inpt = deepcopy(inpt)
```
Comment (josh146) [inline pennylane/templates/utils.py]: How come the deep copy is required?

File: tests/test_templates.py [hunk 2]
```diff
@@ -22,12 +22,12 @@
 * When adding a new interface, try to import it and extend the fixture ``interfaces``. Also add the interface
   gradient computation to the TestGradientIntegration tests.
 
-* When adding a new template, extend the fixtures ``QUBIT_CONSTANT_INPUT`` or ``CV_CONSTANT_INPUT``
+* When adding a new template, extend the fixtures ``QUBIT_DIFFABLE_NONDIFFABLE`` or ``CV_DIFFABLE_NONDIFFABLE``
   by a *list* of arguments to the
   template. Note: Even if the template takes only one argument, it has to be wrapped in a list (i.e. [weights]).
```
Comment (josh146) [inline tests/test_templates.py]: These second sentences also need to be updated right? I.e., since we're now passing a dict

File: tests/test_templates.py [hunk 3]
```diff
@@ -87,116 +107,151 @@
...
+                             'phi_1': [[3.47], [2.01]],
+                             'varphi_1': [[0.93, 1.58], [5.07, 4.82]],
+                             'r': [[0.21, 0.12], [-0.09, 0.04]],
+                             'phi_r': [[4.76, 6.08], [6.09, 6.22]],
+                             'theta_2': [[4.83], [1.70]],
+                             'phi_2': [[4.74], [5.39]],
+                             'varphi_2': [[0.88, 0.62], [1.09, 3.02]],
+                             'a': [[-0.01, -0.05], [0.08, -0.19]],
+                             'phi_a': [[1.89, 3.59], [1.49, 3.71]],
+                             'k': [[0.09, 0.03], [-0.14, 0.04]]},
```
Comment (josh146) [inline tests/test_templates.py]: This is much easier to read :+1:

File: tests/test_templates.py [hunk 4]
```diff
@@ -87,116 +107,151 @@
...
-        inp2 = inp[n:]
-        # Circuit
-        qml.PauliX(wires=0)
-        templ1(*inp1, **hyperp1)
-        templ2(*inp2, **hyperp2)
-        qml.PauliX(wires=1)
-        return [qml.expval(qml.Identity(0)), qml.expval(qml.PauliX(1))]
+    # Signature juggling to pass diffable as single positional arg, but keep track of input names
+    @qml.qnode(dev, interface=interface)
+    def circuit(*diffable, keys_diffable1=None, keys_diffable2=None, nondiffable1=None, nondiffable2=None):
```
Comment (josh146) [inline tests/test_templates.py]: I'm a bit confused as to what these different arguments should correspond to?

File: tests/test_templates.py (hunk 2 above)
Comment (mariaschuld) [reply tests/test_templates.py]: Good find

File: pennylane/templates/utils.py (hunk 1 above)
Comment (mariaschuld) [reply pennylane/templates/utils.py]: You are right, I cannot think of a situation where the function actually changes an object in python...deleted it

File: tests/test_templates.py (hunk 4 above)
Comment (mariaschuld) [reply tests/test_templates.py]: Updated the comment a bit. All that clutter just there to pass the values of the ``diffable`` dictionary, but keep track of the keys to turn it back into a dictionary.

**Comment by mariaschuld** [comment]
One of those odd coverage changes... Although locally coverage of templates is 100% throughout.

Screenshot from 2020-02-16 10-21-46

Is it possible that the reason is the failing "about" test?

**Comment by josh146** [comment]
> One of those odd coverage changes... Although locally coverage of templates is 100% throughout.

Seems that all tests are passing on Travis! Note that the codecov/patch is not a required check, so the PR can be merged without it passing.
